walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,81 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Using linked list as removing the first element will take O(1) time
+    LinkedList<Event> eventList = new LinkedList<Event>(events);
+    // Collections.sort() -> O(nlogn)
+    Collections.sort(eventList, Event.ORDER_EVENT_BY_START);
+
+    // making sure the request is within bounds
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<TimeRange>();
+    }
+
+    if (request.getDuration() < 0) {
+      return new ArrayList<TimeRange>();
+    }
+
+    return traverseThroughSchedule(request, eventList);
+  }
+  /**
+   * Traverses through the schedule and returns the list of available times
+   *
+   * @param request The requested meeting time duration.
+   * @param eventList A sorted list of events ascending from the starting times.
+   * @return Array of open TimeRange objects.
+   */
+  public Collection<TimeRange> traverseThroughSchedule(
+      MeetingRequest request, LinkedList<Event> eventList) {
+
+    ArrayList<TimeRange> openRanges = new ArrayList<TimeRange>();
+
+    long currentTime = (long) TimeRange.START_OF_DAY;
+
+    while (currentTime < (long) TimeRange.END_OF_DAY) {
+
+      // if the last open time slot is between the last meeting and the end of the day
```

Comment:
nit: Remove the "if". This is declarative and describes what the chunk of code does rather than when it runs.